### ISSUE TITLE
[BUG] PnP PowerShell Azure Function. Connect-PnPOnline Fails

### ISSUE THREAD

**Issue by BaronSparky** [pnp/powershell#192]
### Reporting an Issue or Missing Feature
Issue

### Expected behavior 
To be able to connect to SPO from a PowerShell Azure function when using the AZ modules also.  The function app has functions needs to interact with both the AZ and PnP.PowerShell modules.

### Actual behavior 
The Connect-PnPOnline cmdlet works fine as long as the "profile.ps1" has the following commented out
```
if ($env:MSI_SECRET) {
    Disable-AzContextAutosave -Scope Process | Out-Null
    Connect-AzAccount -Identity
}
```
with the above lines apparent in the "profile.ps1", I get:
```
2021-01-28T08:56:50 Welcome, you are now connected to log-streaming service. The default timeout is 2 hours. Change the timeout with the App Setting SCM_LOGSTREAM_TIMEOUT (in seconds).
2021-01-28T08:56:56.750 [Information] Executing 'Functions.TestTrigger' (Reason='This function was programmatically called via the host APIs.', Id=9ce6f59b-0ca5-469e-978b-a8911311841f)
2021-01-28T08:57:12.404 [Information] OUTPUT:
2021-01-28T08:57:13.076 [Information] OUTPUT: Account SubscriptionName TenantId Environment
2021-01-28T08:57:13.077 [Information] OUTPUT: ------- ---------------- -------- -----------
2021-01-28T08:57:13.079 [Information] OUTPUT: ...... Visual Studio Ultimate ..... AzureCloud
2021-01-28T08:57:13.080 [Information] OUTPUT:
2021-01-28T08:57:13.150 [Information] INFORMATION: PowerShell HTTP trigger function processed a request.
2021-01-28T08:57:25.483 [Error] ERROR: The type initializer for 'Microsoft.ApplicationInsights.Extensibility.Implementation.Tracing.CoreEventSource' threw an exception.Exception :Type : System.TypeInitializationExceptionTypeName : Microsoft.ApplicationInsights.Extensibility.Implementation.Tracing.CoreEventSourceTargetSite :Name : <.ctor>b__6_0DeclaringType : Microsoft.ApplicationInsights.Channel.InMemoryTransmitter+<>c, Microsoft.ApplicationInsights, Version=2.16.0.18277, Culture=neutral, PublicKeyToken=31bf3856ad364e35MemberType : MethodModule : Microsoft.ApplicationInsights.dllStackTrace :at Microsoft.ApplicationInsights.Channel.InMemoryTransmitter.<>c.<.ctor>b__6_0(Task task)at System.Threading.Tasks.ContinuationTaskFromTask.InnerInvoke()at System.Threading.Tasks.Task.<>c.<.cctor>b__274_0(Object obj)at System.Threading.ExecutionContext.RunFromThreadPoolDispatchLoop(Thread threadPoolThread, ExecutionContext executionContext, ContextCallback callback, Object state)--- End of stack trace from previous location where exception was thrown ---at System.Threading.ExecutionContext.RunFromThreadPoolDispatchLoop(Thread threadPoolThread, ExecutionContext executionContext, ContextCallback callback, Object state)at System.Threading.Tasks.Task.ExecuteWithThreadLocal(Task& currentTaskSlot, Thread threadPoolThread)Message : The type initializer for 'Microsoft.ApplicationInsights.Extensibility.Implementation.Tracing.CoreEventSource' threw an exception.Data : System.Collections.ListDictionaryInternalInnerException :Type : System.NullReferenceExceptionTargetSite :Name : OnEventSourceCommandDeclaringType : System.Diagnostics.Tracing.CounterGroup, System.Private.CoreLib, Version=4.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798eMemberType : MethodModule : System.Private.CoreLib.dllStackTrace :at System.Diagnostics.Tracing.CounterGroup.OnEventSourceCommand(Object sender, EventCommandEventArgs e)at System.Diagnostics.Tracing.EventSource.add_EventCommandExecuted(EventHandler`1 value)at System.Diagnostics.Tracing.CounterGroup.RegisterCommandCallback()at System.Diagnostics.Tracing.CounterGroup..ctor(EventSource eventSource)at System.Diagnostics.Tracing.CounterGroup.GetCounterGroup(EventSource eventSource)at System.Diagnostics.Tracing.DiagnosticCounter..ctor(String name, EventSource eventSource)at System.Diagnostics.Tracing.EventCounter..ctor(String name, EventSource eventSource)at Microsoft.ApplicationInsights.Extensibility.Implementation.Tracing.CoreEventSource..ctor()at Microsoft.ApplicationInsights.Extensibility.Implementation.Tracing.CoreEventSource..cctor()Message : Object reference not set to an instance of an object.Source : System.Private.CoreLibHResult : -2147467261Source : Microsoft.ApplicationInsightsHResult : -2146233036CategoryInfo : NotSpecified: (:) [Connect-PnPOnline], TypeInitializationExceptionFullyQualifiedErrorId : System.TypeInitializationException,PnP.PowerShell.Commands.Base.ConnectOnlineInvocationInfo :MyCommand : Connect-PnPOnlineScriptLineNumber : 5OffsetInLine : 5HistoryId : 1Line : Connect-PnPOnline -Url https://tenant.sharepoint.com/sites/Provisioning -Credentials $credentialsPositionMessage : At line:5 char:5+ Connect-PnPOnline -Url https://tenant.sharepoint.com/sites/P …+ ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~InvocationName : Connect-PnPOnlineCommandOrigin : InternalScriptStackTrace : at <ScriptBlock>, <No file>: line 5Microsoft.Azure.WebJobs.Script.Workers.Rpc.RpcException : Result: ERROR: The type initializer for 'Microsoft.ApplicationInsights.Extensibility.Implementation.Tracing.CoreEventSource' threw an exception.Exception :Type : System.TypeInitializationExceptionTypeName : Microsoft.ApplicationInsights.Extensibility.Implementation.Tracing.CoreEventSourceTargetSite :Name : <.ctor>b__6_0DeclaringType : Microsoft.ApplicationInsights.Channel.InMemoryTransmitter+<>c, Microsoft.ApplicationInsights, Version=2.16.0.18277, Culture=neutral, PublicKeyToken=31bf3856ad364e35MemberType : MethodModule : Microsoft.ApplicationInsights.dllStackTrace :at Microsoft.ApplicationInsights.Channel.InMemoryTransmitter.<>c.<.ctor>b__6_0(Task task)at System.Threading.Tasks.ContinuationTaskFromTask.InnerInvoke()at System.Threading.Tasks.Task.<>c.<.cctor>b__274_0(Object obj)at System.Threading.ExecutionContext.RunFromThreadPoolDispatchLoop(Thread threadPoolThread, ExecutionContext executionContext, ContextCallback callback, Object state)--- End of stack trace from previous location where exception was thrown ---at System.Threading.ExecutionContext.RunFromThreadPoolDispatchLoop(Thread threadPoolThread, ExecutionContext executionContext, ContextCallback callback, Object state)at System.Threading.Tasks.Task.ExecuteWithThreadLocal(Task& currentTaskSlot, Thread threadPoolThread)Message : The type initializer for 'Microsoft.ApplicationInsights.Extensibility.Implementation.Tracing.CoreEventSource' threw an exception.Data : System.Collections.ListDictionaryInternalInnerException :Type : System.NullReferenceExceptionTargetSite :Name : OnEventSourceCommandDeclaringType : System.Diagnostics.Tracing.CounterGroup, System.Private.CoreLib, Version=4.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798eMemberType : MethodModule : System.Private.CoreLib.dllStackTrace :at System.Diagnostics.Tracing.CounterGroup.OnEventSourceCommand(Object sender, EventCommandEventArgs e)at System.Diagnostics.Tracing.EventSource.add_EventCommandExecuted(EventHandler`1 value)at System.Diagnostics.Tracing.CounterGroup.RegisterCommandCallback()at System.Diagnostics.Tracing.CounterGroup..ctor(EventSource eventSource)at System.Diagnostics.Tracing.CounterGroup.GetCounterGroup(EventSource eventSource)at System.Diagnostics.Tracing.DiagnosticCounter..ctor(String name, EventSource eventSource)at System.Diagnostics.Tracing.EventCounter..ctor(String name, EventSource eventSource)at Microsoft.ApplicationInsights.Extensibility.Implementation.Tracing.CoreEventSource..ctor()at Microsoft.ApplicationInsights.Extensibility.Implementation.Tracing.CoreEventSource..cctor()Message : Object reference not set to an instance of an object.Source : System.Private.CoreLibHResult : -2147467261Source : Microsoft.ApplicationInsightsHResult : -2146233036CategoryInfo : NotSpecified: (:) [Connect-PnPOnline], TypeInitializationExceptionFullyQualifiedErrorId : System.TypeInitializationException,PnP.PowerShell.Commands.Base.ConnectOnlineInvocationInfo :MyCommand : Connect-PnPOnlineScriptLineNumber : 5OffsetInLine : 5HistoryId : 1Line : Connect-PnPOnline -Url https://tenant.sharepoint.com/sites/Provisioning -Credentials $credentialsPositionMessage : At line:5 char:5+ Connect-PnPOnline -Url https://tenant.sharepoint.com/sites/P …+ ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~InvocationName : Connect-PnPOnlineCommandOrigin : InternalScriptStackTrace : at <ScriptBlock>, <No file>: line 5Exception: The type initializer for 'Microsoft.ApplicationInsights.Extensibility.Implementation.Tracing.CoreEventSource' threw an exception.Stack: at Microsoft.ApplicationInsights.Channel.InMemoryTransmitter.<>c.<.ctor>b__6_0(Task task)at System.Threading.Tasks.ContinuationTaskFromTask.InnerInvoke()at System.Threading.Tasks.Task.<>c.<.cctor>b__274_0(Object obj)at System.Threading.ExecutionContext.RunFromThreadPoolDispatchLoop(Thread threadPoolThread, ExecutionContext executionContext, ContextCallback callback, Object state)--- End of stack trace from previous location where exception was thrown ---at System.Threading.ExecutionContext.RunFromThreadPoolDispatchLoop(Thread threadPoolThread, ExecutionContext executionContext, ContextCallback callback, Object state)at System.Threading.Tasks.Task.ExecuteWithThreadLocal(Task& currentTaskSlot, Thread threadPoolThread)
2021-01-28T08:57:25.779 [Error] ERROR: There is currently no connection yet. Use Connect-PnPOnline to connect.Exception :Type : System.InvalidOperationExceptionTargetSite :Name : BeginProcessingDeclaringType : PnP.PowerShell.Commands.Base.PnPConnectedCmdletMemberType : MethodModule : PnP.PowerShell.dllStackTrace :at PnP.PowerShell.Commands.Base.PnPConnectedCmdlet.BeginProcessing()at PnP.PowerShell.Commands.PnPSharePointCmdlet.BeginProcessing()at System.Management.Automation.Cmdlet.DoBeginProcessing()at System.Management.Automation.CommandProcessorBase.DoBegin()Message : There is currently no connection yet. Use Connect-PnPOnline to connect.Source : PnP.PowerShellHResult : -2146233079CategoryInfo : NotSpecified: (:) [Get-PnPWeb], InvalidOperationExceptionFullyQualifiedErrorId : System.InvalidOperationException,PnP.PowerShell.Commands.GetWebInvocationInfo :MyCommand : Get-PnPWebScriptLineNumber : 7OffsetInLine : 12HistoryId : 1Line : $web = Get-PnPWeb;PositionMessage : At line:7 char:12+ $web = Get-PnPWeb;+ ~~~~~~~~~~InvocationName : Get-PnPWebCommandOrigin : InternalScriptStackTrace : at <ScriptBlock>, <No file>: line 7Microsoft.Azure.WebJobs.Script.Workers.Rpc.RpcException : Result: ERROR: There is currently no connection yet. Use Connect-PnPOnline to connect.Exception :Type : System.InvalidOperationExceptionTargetSite :Name : BeginProcessingDeclaringType : PnP.PowerShell.Commands.Base.PnPConnectedCmdletMemberType : MethodModule : PnP.PowerShell.dllStackTrace :at PnP.PowerShell.Commands.Base.PnPConnectedCmdlet.BeginProcessing()at PnP.PowerShell.Commands.PnPSharePointCmdlet.BeginProcessing()at System.Management.Automation.Cmdlet.DoBeginProcessing()at System.Management.Automation.CommandProcessorBase.DoBegin()Message : There is currently no connection yet. Use Connect-PnPOnline to connect.Source : PnP.PowerShellHResult : -2146233079CategoryInfo : NotSpecified: (:) [Get-PnPWeb], InvalidOperationExceptionFullyQualifiedErrorId : System.InvalidOperationException,PnP.PowerShell.Commands.GetWebInvocationInfo :MyCommand : Get-PnPWebScriptLineNumber : 7OffsetInLine : 12HistoryId : 1Line : $web = Get-PnPWeb;PositionMessage : At line:7 char:12+ $web = Get-PnPWeb;+ ~~~~~~~~~~InvocationName : Get-PnPWebCommandOrigin : InternalScriptStackTrace : at <ScriptBlock>, <No file>: line 7Exception: There is currently no connection yet. Use Connect-PnPOnline to connect.Stack: at PnP.PowerShell.Commands.Base.PnPConnectedCmdlet.BeginProcessing()at PnP.PowerShell.Commands.PnPSharePointCmdlet.BeginProcessing()at System.Management.Automation.Cmdlet.DoBeginProcessing()at System.Management.Automation.CommandProcessorBase.DoBegin()
2021-01-28T08:57:25.814 [Information] Executed 'Functions.TestTrigger' (Succeeded, Id=9ce6f59b-0ca5-469e-978b-a8911311841f, Duration=29065ms)
```

### Steps to reproduce behavior
Using sample code provided:
```
using namespace System.Net

# Input bindings are passed in via param block.
param($Request, $TriggerMetadata)

# Write to the Azure Functions log stream.
Write-Host "PowerShell HTTP trigger function processed a request."

$script = {
    $securePassword = ConvertTo-SecureString $env:pnp_pass -AsPlainText -Force
    $credentials = New-Object PSCredential ($env:pnp_user, $securePassword)

    Connect-PnPOnline -Url https://tenant.sharepoint.com/sites/Provisioning -Credentials $credentials

    $web = Get-PnPWeb;
    $web.Title
}

$webTitle = Start-ThreadJob -Script $script | Receive-Job -Wait

$body = "The title of the web is $($webTitle)"

# Associate values to output bindings by calling 'Push-OutputBinding'.
Push-OutputBinding -Name Response -Value ([HttpResponseContext]@{
        StatusCode = [HttpStatusCode]::OK
        Body = $body
    })
```

Requirements.psd1
```
# This file enables modules to be automatically managed by the Functions service.
# See https://aka.ms/functionsmanageddependency for additional information.
@{
    'Az' = '5.*'
    'AzTable' = '2.*'
    'Az.Accounts' = '2.2.0'
    'Az.Storage' = '3.1.0'
    'PnP.PowerShell' = '1.*'
}
```

Profiles.ps1
```
# Azure Functions profile.ps1
#
# This profile.ps1 will get executed every "cold start" of your Function App.
# "cold start" occurs when:
#
# * A Function App starts up for the very first time
# * A Function App starts up after being de-allocated due to inactivity
#
# You can define helper functions, run commands, or specify environment variables
# NOTE: any variables defined that are not environment variables will get reset after the first execution

# Authenticate with Azure PowerShell using MSI.
# Remove this if you are not planning on using MSI or Azure PowerShell.
if ($env:MSI_SECRET) {
    Disable-AzContextAutosave -Scope Process | Out-Null
    Connect-AzAccount -Identity
}

# Uncomment the next line to enable legacy AzureRm alias in Azure PowerShell.
# Enable-AzureRmAlias

# You can also define functions or aliases that can be referenced in any of your PowerShell functions.
```

### What is the version of the Cmdlet module you are running?
1.2.0

### Which operating system/environment are you running PnP PowerShell on?
- [ ] Windows
- [ ] Linux
- [ ] MacOS
- [ ] Azure Cloud Shell
- [X] Azure Functions
- [ ] Other : please specify

**Comment by erwinvanhunen** [comment]
The reason for this issue is that the Az modules use a different version of the application insights assemblies than what PnP PowerShell uses. This is a general problem in PowerShell and not specific to the Az or PnP PowerShell modules. Once an assembly has been loaded into the appdomain then one can't load another version.  Effectively we are loading the appinsights assembly in a different context already (assembly load context), especially for this reason and we'll look into why this does not seem to work anymore.

**Comment by erwinvanhunen** [comment]
Actually, having said that, I just tried exactly what you're doing and have no issues...

This is what I have in my requirements.psd1

```powershell
@{
    # For latest supported version, go to 'https://www.powershellgallery.com/packages/Az'. 
    'Az' = '5.*'
    'AzTable' = '2.*'
    'Az.Accounts' = '2.2.0'
    'Az.Storage' = '3.1.0'
    'PnP.PowerShell' = '1.*'
}
```

And this is my profile.ps1

```powershell
if ($env:MSI_SECRET) {
   Disable-AzContextAutosave -Scope Process | Out-Null
   Connect-AzAccount -Identity
}
```

and this is the function code:

```powershell
using namespace System.Net

# Input bindings are passed in via param block.
param($Request, $TriggerMetadata)

# Write to the Azure Functions log stream.
Write-Host "PowerShell HTTP trigger function processed a request."

$scriptBlock = {
    [securestring]$secStringPassword = ConvertTo-SecureString $env:pnp_pass -AsPlainText -Force
    [pscredential]$credObject = New-Object System.Management.Automation.PSCredential ($env:pnp_user, $secStringPassword)
    
    Connect-PnPOnline -Url https://erwinmcm.sharepoint.com/sites/demo1 -Credentials $credObject
    $web = Get-PnPWeb
    $web.Title
}

$webTitle = Start-ThreadJob -Script $scriptBlock | Receive-Job -Wait

$body = "Title $webTitle"

# Associate values to output bindings by calling 'Push-OutputBinding'.
Push-OutputBinding -Name Response -Value ([HttpResponseContext]@{
    StatusCode = [HttpStatusCode]::OK
    Body = $body
})
```

E.g. a full copy of your settings. And I have no issue running the function:

![image](https://user-images.githubusercontent.com/7666381/106134249-90dea180-6166-11eb-8ef7-ab2ae46245d5.png)

**Comment by erwinvanhunen** [comment]
Have you tried stopping and starting the function app? As that will clear the module cache, and redownload the latest versions according to the requirements.psd1. Maybe you have a different version of the modules loaded?

These are the modules on my end (I modifed the function output the results of `Get-Module -ListAvailable | Select Name,Version`)

Az: 5.4.0
AzTable: 2.0.3
Az.Accounts: 2.2.4 *and* 2.2.0 (confusingly enough)
Az.Storage: 3.2.1 *and* 3.1.0
PnP.PowerShell: 1.2.0

**Comment by BaronSparky** [comment]
Hi Erwin,

Thanks for the quick reply.

I have the same set of modules as you, including the multiples.

Thinking back, I never saw the message:
`2021-01-28T13:54:43.020 [Warning] The first managed dependency download is in progress, function execution will continue when it's done. Depending on the content of requirements.psd1, this can take a few minutes. Subsequent function executions will not block and updates will be performed in the background.`

after updating for PnP.PowerShell 1.2.0 from 0.3.35.

Tried restarting the Function app and stopping/starting with no change.

However, when I created a new Function App in the same resource group, I saw the above message regarding the downloading of modules, and now all seem to be running as expected.

Many thanks.
